addOns/paramminer/src/main/java/org/zaproxy/addon/paramminer/ParamGuesserScanController.java

```diff
@@ -0,0 +1,254 @@
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.zaproxy.addon.paramminer.gui.ParamMinerPanel;
+import org.zaproxy.zap.model.ScanController;
+import org.zaproxy.zap.model.Target;
+import org.zaproxy.zap.users.User;
+
+public class ParamGuesserScanController implements ScanController<GuesserScan> {
+    private final Lock paramGuesserScanLock;
+    private int scanIdCounter;
+    private Map<Integer, GuesserScan> paramGuesserScanMap;
+
+    private ParamMinerPanel panel;
+
+    public ParamGuesserScanController() {
+        this.paramGuesserScanLock = new ReentrantLock();
+        this.scanIdCounter = 0;
+        this.paramGuesserScanMap = new HashMap<>();
+    }
+
+    public int startScan(String displayName, ParamMinerConfig config) {
+        paramGuesserScanLock.lock();
+        try {
+            int id = this.scanIdCounter++;
+            GuesserScan paramGuesserScan = new GuesserScan(id, config, displayName);
+            paramGuesserScanMap.put(id, paramGuesserScan);
+            paramGuesserScan.start();
+
+            if (panel != null) {
+                panel.scannerStarted(paramGuesserScan);
+                panel.setTabFocus();
+            }
+            return id;
+        } finally {
+            paramGuesserScanLock.unlock();
+        }
+    }
+
+    @Override
+    public int startScan(
+            String displayName, Target target, User user, Object[] contextSpecificObjects) {
+        throw new UnsupportedOperationException(
+                "Scans are started with a param miner configuration.");
+    }
+
+    @Override
+    public List<GuesserScan> getAllScans() {
+        paramGuesserScanLock.lock();
+        try {
+            return new ArrayList<>(paramGuesserScanMap.values());
+        } finally {
+            paramGuesserScanLock.unlock();
+        }
+    }
+
+    @Override
+    public List<GuesserScan> getActiveScans() {
+        List<GuesserScan> scans = new ArrayList<>();
+        paramGuesserScanLock.lock();
+        try {
+            for (GuesserScan scan : paramGuesserScanMap.values()) {
+                if (!scan.isStopped()) {
+                    scans.add(scan);
+                }
+            }
+            return scans;
+        } finally {
+            paramGuesserScanLock.unlock();
+        }
+    }
+
+    @Override
+    public GuesserScan getScan(int id) {
+        return this.paramGuesserScanMap.get(id);
+    }
+
+    @Override
+    public void stopScan(int id) {
+        paramGuesserScanLock.lock();
+        try {
+            if (this.paramGuesserScanMap.containsKey(id)) {
+                this.paramGuesserScanMap.get(id).stopScan();
+                updateScansPanel();
+            }
+        } finally {
+            paramGuesserScanLock.unlock();
+        }
+    }
+
+    @Override
+    public void stopAllScans() {
+        paramGuesserScanLock.lock();
+        try {
+            paramGuesserScanMap.values().forEach(GuesserScan::resumeScan);
```

Comment:
I'm confused, resume on stop?

Reply:
Oops :)

Reply:
Try to resume it and then stop it maybe .. if stop fails for some reason

Reply:
Wait ... where is it stopping?

Reply:
Should be stopScan

Reply:
changed it to stopScan :)